File: app/src/androidTest/java/com/bopr/android/smailer/CallReceiverTest.java
package com.bopr.android.smailer;

import android.content.Context;
import android.content.Intent;

import org.junit.Test;

import static android.content.Intent.ACTION_NEW_OUTGOING_CALL;
import static android.telephony.SmsMessage.FORMAT_3GPP;
import static android.telephony.TelephonyManager.ACTION_PHONE_STATE_CHANGED;
import static android.telephony.TelephonyManager.EXTRA_INCOMING_NUMBER;
import static android.telephony.TelephonyManager.EXTRA_STATE;
import static android.telephony.TelephonyManager.EXTRA_STATE_IDLE;
import static android.telephony.TelephonyManager.EXTRA_STATE_OFFHOOK;
import static android.telephony.TelephonyManager.EXTRA_STATE_RINGING;
import static com.bopr.android.smailer.CallReceiver.SMS_RECEIVED;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * {@link CallReceiver} tester.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class CallReceiverTest extends BaseTest {

    private Context context;

    @Override
    public void setUp() throws Exception {
        super.setUp();

        context = mock(Context.class);
        when(context.getResources()).thenReturn(getContext().getResources());
    }

    /**
     * Checks that receiver starts service on incoming call.
     */
    @Test
    public void testReceiveIncomingCall() {
        MethodInvocationsCollector invocations = new MethodInvocationsCollector();

        doAnswer(invocations).when(context).startService(any(Intent.class));

        CallReceiver receiver = new CallReceiver();

        /* ringing */
        Intent intent = new Intent(ACTION_PHONE_STATE_CHANGED);
        intent.putExtra(EXTRA_STATE, EXTRA_STATE_RINGING);
        intent.putExtra(EXTRA_INCOMING_NUMBER, "123");

        receiver.onReceive(context, intent);

        assertTrue(invocations.isEmpty());

        /* off hook */
        intent = new Intent(ACTION_PHONE_STATE_CHANGED);
        intent.putExtra(EXTRA_STATE, EXTRA_STATE_OFFHOOK);

        receiver.onReceive(context, intent);

        assertTrue(invocations.isEmpty());

        /* end call */
        intent = new Intent(ACTION_PHONE_STATE_CHANGED);
        intent.putExtra(EXTRA_STATE, EXTRA_STATE_IDLE);

        receiver.onReceive(context, intent);

        Intent result = invocations.getArgument(0, 0);
        PhoneEvent event = result.getParcelableExtra("event");

        assertNotNull(event);
        assertEquals("123", event.getPhone());
        assertTrue(event.isIncoming());
    }

    /**
     * Checks that receiver starts service on outgoing call.
     */
    @Test
    public void testReceiveOutgoingCall() {
        MethodInvocationsCollector invocations = new MethodInvocationsCollector();
        doAnswer(invocations).when(context).startService(any(Intent.class));

        CallReceiver receiver = new CallReceiver();

        /* ringing */
        Intent intent = new Intent(ACTION_NEW_OUTGOING_CALL);
        intent.putExtra(Intent.EXTRA_PHONE_NUMBER, "123");

        receiver.onReceive(context, intent);

        assertTrue(invocations.isEmpty());

        /* off hook */
        intent = new Intent(ACTION_PHONE_STATE_CHANGED);
        intent.putExtra(EXTRA_STATE, EXTRA_STATE_OFFHOOK);

        receiver.onReceive(context, intent);

        assertTrue(invocations.isEmpty());

        /* end call */
        intent = new Intent(ACTION_PHONE_STATE_CHANGED);
        intent.putExtra(EXTRA_STATE, EXTRA_STATE_IDLE);

        receiver.onReceive(context, intent);

        Intent result = invocations.getArgument(0, 0);
        PhoneEvent event = result.getParcelableExtra("event");

        assertNotNull(event);
        assertEquals("123", event.getPhone());
        assertFalse(event.isIncoming());
    }

    /**
     * Checks that receiver starts service on missed call.
     */
    @Test
    public void testReceiveMissedCall() {
        MethodInvocationsCollector invocations = new MethodInvocationsCollector();
        doAnswer(invocations).when(context).startService(any(Intent.class));

        CallReceiver receiver = new CallReceiver();

        /* ringing */
        Intent intent = new Intent(ACTION_PHONE_STATE_CHANGED);
        intent.putExtra(EXTRA_STATE, EXTRA_STATE_RINGING);
        intent.putExtra(EXTRA_INCOMING_NUMBER, "123");

        receiver.onReceive(context, intent);

        assertTrue(invocations.isEmpty());

        /* end call */
        intent = new Intent(ACTION_PHONE_STATE_CHANGED);
        intent.putExtra(EXTRA_STATE, EXTRA_STATE_IDLE);

        receiver.onReceive(context, intent);

        Intent result = invocations.getArgument(0, 0);
        PhoneEvent event = result.getParcelableExtra("event");

        assertNotNull(event);
        assertEquals("123", event.getPhone());
        assertTrue(event.isMissed());
    }

    /**
     * Checks that receiver starts service on sms.
     */
    @Test
    public void testReceiveSms() {
        MethodInvocationsCollector invocations = new MethodInvocationsCollector();
        doAnswer(invocations).when(context).startService(any(Intent.class));

        CallReceiver receiver = new CallReceiver();

        Intent intent = new Intent(SMS_RECEIVED);
        intent.putExtra("format", FORMAT_3GPP);
        intent.putExtra("pdus", new Object[]{new byte[]{0, 32, 11, -111, 81, 85, 37, 81, 85, -10, 0,
                0, 2, 16, 98, 2, 16, -109, 41, 12, -44, 50, -98, 14, 106, -105, -25, -13, -16, -71, 12}});

        receiver.onReceive(context, intent);

        Intent result = invocations.getArgument(0, 0);
        PhoneEvent event = result.getParcelableExtra("event");

        assertNotNull(event);
        assertEquals("+15555215556", event.getPhone());
        assertEquals("Text message", event.getText());
    }

}

File: app/src/androidTest/kotlin/com/bopr/android/smailer/CallReceiverTest.kt
package com.bopr.android.smailer

import android.Manifest.permission.READ_CALL_LOG
import android.Manifest.permission.READ_PHONE_STATE
import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.os.Build
import android.telephony.SmsMessage
import android.telephony.TelephonyManager.*
import androidx.test.filters.SmallTest
import androidx.test.rule.GrantPermissionRule
import androidx.test.rule.GrantPermissionRule.grant
import com.bopr.android.smailer.PhoneEvent.Companion.STATE_PENDING
import com.bopr.android.smailer.util.deviceName
import com.nhaarman.mockitokotlin2.*
import org.junit.Before
import org.junit.Rule
import org.junit.Test

/**
 * [CallReceiver] tester.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
@SmallTest
class CallReceiverTest : BaseTest() {

    @get:Rule
    val permissionRule: GrantPermissionRule = grant(READ_CALL_LOG, READ_PHONE_STATE)

    private lateinit var context: Context

    @Before
    fun setUp() {
        context = mock()
    }

    @Suppress("DEPRECATION")
    private fun calStateIntent(state: String, number: String?): Intent {
        val intent = Intent(ACTION_PHONE_STATE_CHANGED)
        intent.putExtra(EXTRA_STATE, state)
        number?.run { intent.putExtra(EXTRA_INCOMING_NUMBER, number) }
        return intent
    }

    /**
     * Checks that receiver starts service on incoming call.
     */
    @Test
    fun testReceiveIncomingCall() {
        val receiver = CallReceiver()

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_RINGING, "100"))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_OFFHOOK, "100"))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_IDLE, "100"))
        verify(context).startService(argThat {
            getParcelableExtra<PhoneEvent>("event")!!.run {
                isIncoming && !isMissed && phone == "100"
                        && location == null && endTime != null && acceptor == deviceName()
                        && state == STATE_PENDING
            }
        })
    }

    @Test
    @TargetApi(Build.VERSION_CODES.Q)
    fun testReceiveIncomingCallQ() {
        val receiver = CallReceiver()

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_RINGING, null))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_RINGING, "100"))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_OFFHOOK, null))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_OFFHOOK, "100"))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_IDLE, null))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_IDLE, "100"))
        verify(context).startService(argThat {
            getParcelableExtra<PhoneEvent>("event")!!.run {
                isIncoming && !isMissed && phone == "100"
                        && location == null && endTime != null && acceptor == deviceName()
                        && state == STATE_PENDING
            }
        })
    }

    /**
     * Checks that receiver starts service on outgoing call.
     */
    @Test
    fun testReceiveOutgoingCall() {
        val receiver = CallReceiver()

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_OFFHOOK, "200"))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_IDLE, "200"))
        verify(context).startService(argThat {
            getParcelableExtra<PhoneEvent>("event")!!.run {
                !isIncoming && !isMissed && phone == "200"
                        && location == null && endTime != null && acceptor == deviceName()
                        && state == STATE_PENDING
            }
        })
    }

    /**
     * Checks that receiver starts service on missed call.
     */
    @Test
    fun testReceiveMissedCall() {
        val receiver = CallReceiver()

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_RINGING, "300"))
        verify(context, never()).startService(anyOrNull())

        receiver.onReceive(context, calStateIntent(EXTRA_STATE_IDLE, "300"))
        verify(context).startService(argThat {
            getParcelableExtra<PhoneEvent>("event")!!.run {
                isIncoming && isMissed && phone == "300"
                        && location == null && endTime != null && acceptor == deviceName()
                        && state == STATE_PENDING
            }
        })
    }

    /**
     * Checks that receiver starts service on sms.
     */
    @Test
    fun testReceiveSms() {
        val receiver = CallReceiver()

        val intent = Intent(CallReceiver.SMS_RECEIVED)
        intent.putExtra("format", SmsMessage.FORMAT_3GPP)
        intent.putExtra("pdus", arrayOf<Any>(byteArrayOf(0, 32, 11, -111, 81, 85, 37, 81, 85, -10, 0,
                0, 2, 16, 98, 2, 16, -109, 41, 12, -44, 50, -98, 14, 106, -105, -25, -13, -16, -71, 12)))

        receiver.onReceive(context, intent)

        verify(context).startService(argThat {
            getParcelableExtra<PhoneEvent>("event")!!.run {
                isSms && isIncoming && phone == "+15555215556" && text == "Text message"
                        && location == null && endTime != null && acceptor == deviceName()
                        && state == STATE_PENDING
            }
        })
    }
}